pip/pip-370.md

```diff
@@ -0,0 +1,91 @@
+# It is not a dynamic config, the default value is "true" to preserve backward-compatible behavior. 
+createTopicToRemoteClusterForReplication=true
+```
+
+## Design & Implementation Details
+
+### Phase 1: Introduce a flag to disable the replicators to automatically trigger topic creation.
+- If `createTopicToRemoteClusterForReplication` is set to `false`.
+  1. After enabling namespace-level Geo-Replication: the broker will not create topics on the remote cluster automatically when calling `pulsar-admin topics create-partitioned-topic`.
+  2. When enabling topic-level Geo-Replication on a partitioned topic: broker will not create topics on the remote cluster automatically.
+  3. The internal Geo-Replicator will not trigger topic auto-creation for remote clusters, it just keeps retrying to check if the topic exists on the remote cluster, once the topic is created, the replicator starts.
+  4. It does not change the behavior of creating subscriptions after enabling `enableReplicatedSubscriptions`, the subscription will also be created on the remote cluster after users enable. `enableReplicatedSubscriptions`.
+  5. The config `allowAutoTopicCreation` still works for the local cluster as before, it will not be affected by the new config `createTopicToRemoteClusterForReplication`.
+- If `createTopicToRemoteClusterForReplication` is set to `true`.
+  a. All components work as before, see details: `Motivation -> The current topic creation behavior when enabling Geo-Replication`
+
+### Phase 2: The replicator will check remote topics' partitioned metadata and update partitions in the remote cluster to the same as the current cluster if needed.
+- If `createTopicToRemoteClusterForReplication` is set to `false`.
+  - The behavior is the same as Phase 1.
+- If `createTopicToRemoteClusterForReplication` is set to `true`.
+  - When a replicator for a topic partition starts, it checks the partitioned metadata in the remote cluster first and updates partitions in the remote cluster to the same as the current cluster if needed. Seem the example as follows:
+    -  `local_cluster.topic.partitions = 2` and the topic does not exist in the remote cluster: create a partitioned topic with `2` partitions in the remote cluster.
+      - Before `PIP-370 Phase 2`: the replicator will only trigger partition creation (`{topic}-partition-0` and `{topic}-partition-1`), and will not care about partitioned metadata.
+    -  `local_cluster.topic.partitions = 2` and `remote_cluster.topic.partitions = 1`: expand `remote_cluster.topic.partitions` to `2`.
+      - Before `PIP-370 Phase 2`: the replicator will only trigger partition creation (`{topic}-partition-0` and `{topic}-partition-1`), and the partitioned metadata in the remote cluster is still `1`. 
+    -  `local_cluster.topic.partitions = 2` and `remote_cluster.topic.partitions >=2 `: modifies nothing.
```

Comment:
Maybe we'd have a flag in the future to only start the replicator with consistent partitioned metadata across clusters. The reason is the messages with same key published to the partitioned topic from different cluster will go to different partition, that will be a tricky issue for the ordering message processing case.

Reply:
For this scenario, it is fine. The messages will be copied to the same partition in the remote cluster, and no message will be copied to the partition that is larger than `{local_cluster.topic.partitions}` in the remote cluster